### PR TITLE
Multiple secondary NS servers

### PR DESCRIPTION
**Pull request by bbustin**
Added the ability to use more than one secondary NS server. Changes made to the daemon to allow this to work from the web interface.

### CODE REVIEW

**Comment by bbustin** [comment]
In the process of fixing a bug in the domain checks caused by some of the changes in this pull request.

File: management/dns_update.py [hunk 1]
```diff
@@ -144,8 +144,12 @@ def build_zone(domain, all_domains, additional_records, www_redirect_domains, en
 		records.append((None,  "NS",  "ns1.%s." % env["PRIMARY_HOSTNAME"], False))
 
 		# Define ns2.PRIMARY_HOSTNAME or whatever the user overrides.
-		secondary_ns = get_secondary_dns(additional_records) or ("ns2." + env["PRIMARY_HOSTNAME"])
-		records.append((None,  "NS", secondary_ns+'.', False))
+		# User may provide one or more additional nameservers
+		if get_secondary_dns(additional_records):
+				[records.append((None,  "NS", secondary_ns+'.', False)) for secondary_ns in get_secondary_dns(additional_records)]
```
Comment (JoshData) [inline management/dns_update.py]: There is an extra tab on this line.

Comment (JoshData) [inline management/dns_update.py]: Whoa. You put a call inside a generator. No, that's too award. Please replace with a regular for loop.

File: management/dns_update.py [hunk 2]
```diff
@@ -144,8 +144,12 @@ def build_zone(domain, all_domains, additional_records, www_redirect_domains, en
...
+		else:
+			secondary_ns = get_secondary_dns(additional_records) or ()
```
Comment (JoshData) [inline management/dns_update.py]: This line is unnecessary.

File: management/dns_update.py (hunk 1 above)
Comment (JoshData) [inline management/dns_update.py]: Also please avoid calling get_secondary_dns multiple times here.

File: management/dns_update.py [hunk 3]
```diff
@@ -785,33 +795,38 @@ def set_custom_dns_record(qname, rtype, value, action, env):
...
+		raise ValueError("Valid types are one or more of the following: %s" % ", ".join(valid_types))
+
+	values = []
 	for qname, rtype, value in custom_dns:
-		if qname == "_secondary_nameserver":
-			return value
+		if qname in dns_type:
+			if isinstance(value, str):
+				values.append(value)
+	if len(values) > 0:
```
Comment (JoshData) [inline management/dns_update.py]: You can avoid some if statements elsewhere by omitting this and always returning the values list, even if it's empty.

File: management/dns_update.py [hunk 4]
```diff
@@ -903,5 +918,5 @@ def build_recommended_dns(env):
 		for zone, records in build_recommended_dns(env):
 			for record in records:
 				print("; " + record['explanation'])
-				print(record['qname'], record['rtype'], record['value'], sep="\t")
+				print(record['qname'], record['rtype'], record['value'])
```
Comment (JoshData) [inline management/dns_update.py]: Unnecessary change.

File: management/dns_update.py [hunk 5]
```diff
@@ -785,33 +795,38 @@ def set_custom_dns_record(qname, rtype, value, action, env):
...
-
-		# Set.
-		set_custom_dns_record("_secondary_nameserver", "A", hostname, "set", env)
+		for item in hostnames:
+			try:
+				response = dns.resolver.query(item, "A")
+			except (dns.resolver.NoNameservers, dns.resolver.NXDOMAIN, dns.resolver.NoAnswer):
+				raise ValueError("Could not resolve the IP address of %s." % item)
+			# Set.
+		set_custom_dns_record("_secondary_nameserver", "A", {"A":hostnames}, "set", env)
```
Comment (JoshData) [inline management/dns_update.py]: This call does not look correct. Try something other than `"set"` here.

Comment (bbustin) [inline management/dns_update.py]: The original was using set, so not sure why it should be different here.

File: management/dns_update.py (hunk 4 above)
Comment (bbustin) [inline management/dns_update.py]: Reverting changs

File: management/dns_update.py (hunk 3 above)
Comment (bbustin) [inline management/dns_update.py]: Making the change.

File: management/dns_update.py (hunk 5 above)
Comment (JoshData) [inline management/dns_update.py]: Because you are setting multiple values. Don't do this `{ "A": ...` thing. That doesn't make sense.

Comment (bbustin) [inline management/dns_update.py]: I don't like that either. Think it is pretty messy. Just could not figure out how to get it to work in an elegant way without a lot of collateral code changing. Probably because I am still not very familiar with the codebase.

Comment (bbustin) [inline management/dns_update.py]: If you have any guidance on a better way to do this, please let me know. I'd love to be able to contribute to the project, even if it is something tiny like multiple secondary nameservers.

Comment (JoshData) [inline management/dns_update.py]: I'm a little swamped trying to dig my way out of the hole I made for myself releasing v0.11, but when I get back to this patch I can look closer.

Comment (bbustin) [inline management/dns_update.py]: Sounds good. Mail-in-a-box is pretty amazing. Thanks for creating this and putting it out there for anyone to use.

Comment (JoshData) [inline management/dns_update.py]: Thanks.

**Comment by JoshData** [comment]
I'm now testing this and will merge shortly. Thanks.

**Comment by bbustin** [comment]
Sounds great.

**Comment by JoshData** [comment]
I squashed your commits together and then tweaked the logic a bit. You can see my changes in the merge commit above. Thanks.

**Comment by bbustin** [comment]
It looks really good. I like how you made it so you could prepend xfr: to
Secondary DNS. There were other parts that make a whole lot more sense now than the way I hacked things up earlier. The flow is much better and it is much more pythonic.
